fastcrypto/src/groups/class_group/mod.rs

```diff
@@ -0,0 +1,394 @@
+            g = xgcd.gcd;
+            let y = xgcd.y;
+            let h = &f / &g;
+            capital_by = u1 / &g;
+            capital_cy = u2 / &g;
+            capital_dy = &s / &g;
+
+            // 4.
+            let l = (&y * (&b * (w1.mod_floor(&h)) + &c * (w2.mod_floor(&h)))).mod_floor(&h);
+            capital_bx = &b * (&m / &h) + &l * (&capital_by / &h);
+        }
+
+        // 5. (partial xgcd)
+        let mut bx = capital_bx.mod_floor(&capital_by);
+        let mut by = capital_by.clone();
+
+        let mut x = BigInt::one();
+        let mut y = BigInt::zero();
+        let mut z = 0u32;
+
+        while by.abs() > self.partial_gcd_limit && !bx.is_zero() {
+            let (q, t) = by.div_rem(&bx);
+            by = bx;
+            bx = t;
+            let tmp = &y - &q * &x;
+            y = x;
+            x = tmp;
+            z += 1;
+        }
+
+        if z.is_odd() {
+            by = -by;
+            y = -y;
+        }
+
+        let (ax, ay) = if g.is_one() {
+            (x.clone(), y.clone())
+        } else {
+            (&g * &x, &g * &y)
+        };
+
+        let u3: BigInt;
+        let w3: BigInt;
+        let v3: BigInt;
+
+        if z == 0 {
+            // 6.
+            let q = &capital_cy * &bx;
+            let cx = (&q - &m) / &capital_by;
+            let dx = (&bx * &capital_dy - w2) / &capital_by;
+            u3 = &by * &capital_cy;
+            w3 = &bx * &cx - &g * &dx;
+            v3 = v2 - (&q << 1);
+        } else {
+            // 7.
+            let cx = (&capital_cy * &bx - &m * &x) / &capital_by;
+            let q1 = &by * &cx;
+            let q2 = &q1 + &m;
+            let dx = (&capital_dy * &bx - w2 * &x) / &capital_by;
+            let q3 = &y * &dx;
+            let q4 = &q3 + &capital_dy;
+            let dy = &q4 / &x;
+            let cy = if !b.is_zero() {
+                &q2 / &bx
+            } else {
+                (&cx * &dy - w1) / &dx
+            };
+            u3 = &by * &cy - &ay * &dy;
+            w3 = &bx * &cx - &ax * &dx;
+            v3 = &g * (&q3 + &q4) - &q1 - &q2;
+        }
+
+        QuadraticForm {
+            a: u3,
+            b: v3,
+            c: w3,
+            partial_gcd_limit: self.partial_gcd_limit.clone(),
+        }
+        .reduce()
+    }
+}
+
+impl ParameterizedGroupElement for QuadraticForm {
+    /// The discriminant of a quadratic form defines the class group.
+    type ParameterType = Discriminant;
+
+    type ScalarType = BigInt;
+
+    fn zero(discriminant: &Self::ParameterType) -> Self {
+        Self::from_a_b_discriminant(BigInt::one(), BigInt::one(), discriminant)
+    }
+
+    fn double(&self) -> Self {
+        self.compose(self)
+    }
+
+    fn mul(&self, scale: &BigInt) -> Self {
+        if scale.is_zero() {
+            return Self::zero(&self.discriminant());
+        }
+
+        let mut result = self.clone();
+        for i in (0..scale.bits() - 1).rev() {
```

Comment:
there is no iterator on the bits?

Reply:
No, there's only the bit length and the check bit functions.